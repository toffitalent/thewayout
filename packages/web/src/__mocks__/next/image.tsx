import * as React from 'react';

// eslint-disable-next-line react/jsx-no-useless-fragment
const mock = (props: { children: React.ReactElement }): React.ReactElement => <>{props.children}</>;

export default mock;
