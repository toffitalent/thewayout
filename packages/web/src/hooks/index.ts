export * from './useAppAuth';
export * from './useAppDispatch';
export * from './useAppSelector';
export * from './useDashboardRedirection';
export * from './useErrorIndicator';
export * from './useFormattedPhoneNumber';
export * from './useIndicators';
export * from './useMediaQuery';
export * from './usePrevious';
export * from './useTimeout';
