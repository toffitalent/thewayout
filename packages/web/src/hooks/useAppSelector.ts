import { TypedUseSelectorHook, useSelector } from 'react-redux';
import type { State } from '@app/store';

export const useAppSelector: TypedUseSelectorHook<State> = useSelector;
