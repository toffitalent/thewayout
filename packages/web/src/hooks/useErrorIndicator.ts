import { Toast } from '@disruptive-labs/ui';
import { useCallback } from 'react';
import { ApiError } from '@app/api';
import { isSerializedApiError, SerializedApiError } from '@app/utils';

const getDefaultMessageByErrorCode = (error: ApiError) => {
  switch (error.code) {
    case 'RequestThrottled':
      return 'Too many requests. Please try again in a few minutes.';
    default:
      return null;
  }
};

const getErrorMessage = (
  error: ApiError | SerializedApiError,
  errorCodeMap: Record<string, string>,
  defaultMessage: string | null,
) => {
  switch (error.type) {
    case ApiError.Types.Client:
      return errorCodeMap?.[error.code] ?? getDefaultMessageByErrorCode(error) ?? defaultMessage;
    case ApiError.Types.Network:
      return 'No connection';
    default:
      return defaultMessage;
  }
};

export const useErrorIndicator = (
  errorCodeMap?: Record<string, string>,
  defaultMessage: string | null = 'Something went wrong',
) =>
  useCallback(
    (error: any, shouldRethrowError = true) => {
      const text =
        error instanceof ApiError || isSerializedApiError(error)
          ? getErrorMessage(error, errorCodeMap ?? {}, defaultMessage)
          : defaultMessage;

      if (text) Toast.show({ status: 'danger', text });

      if (shouldRethrowError) throw error;
    },
    [defaultMessage, errorCodeMap],
  );
