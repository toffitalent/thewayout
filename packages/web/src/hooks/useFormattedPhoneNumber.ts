import type { AsYouType, CountryCode, PhoneNumber } from 'libphonenumber-js';
import { useEffect, useState } from 'react';

type ParsePhoneNumber = (text: string, defaultCountry?: CountryCode | undefined) => PhoneNumber;

const loadModule = async () => {
  const { AsYouType, parsePhoneNumber } = await import('libphonenumber-js');

  return { AsYouType, parsePhoneNumber };
};

export const useFormattedPhoneNumber = () => {
  const [formatNumber, setFormatNumber] = useState<AsYouType>();
  const [parsePhoneNumber, setParsePhoneNumber] = useState<ParsePhoneNumber>();

  useEffect(() => {
    loadModule().then(({ AsYouType, parsePhoneNumber }) => {
      const asYouType = new AsYouType('US');

      setFormatNumber(asYouType);
      setParsePhoneNumber(() => parsePhoneNumber);
    });
  }, []);

  return { formatNumber, parsePhoneNumber };
};
