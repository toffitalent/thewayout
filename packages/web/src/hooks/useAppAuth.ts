import { useAuth } from '@disruptive-labs/client/react';
import { useRouter } from 'next/router';
import { useCallback, useEffect, useState } from 'react';
import { AppClient, UserType } from '@app/api';

export const useAppAuth = () => useAuth<AppClient>();

export function useAuthRedirect(autoRedirectOnFirstRender = true) {
  const auth = useAppAuth();
  const router = useRouter();
  const [shouldRedirect, setShouldRedirect] = useState(false);

  const redirect = useCallback(() => {
    setShouldRedirect(true);
  }, []);

  // Handle auto-redirect on ready, if enabled
  useEffect(() => {
    if (!router.isReady || auth.isLoading) return;

    if (autoRedirectOnFirstRender && auth.isAuthenticated) redirect();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [auth.isLoading, router.isReady, autoRedirectOnFirstRender, redirect]);

  // Handle manual redirect once auth.user available
  useEffect(() => {
    if (shouldRedirect) {
      setShouldRedirect(false);

      const userType = Object.values(UserType).find((item) => auth.user?.roles?.includes(item));
      router.replace(
        userType ? `/${userType !== UserType.Rsp ? userType : `ssp/case-managers`}/` : '/',
      );
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [auth.user, shouldRedirect]);

  return redirect;
}
