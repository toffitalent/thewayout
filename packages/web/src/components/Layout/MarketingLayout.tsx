import { classNames, SkipNavContent, SkipNavLink } from '@disruptive-labs/ui';
import React from 'react';
import { useDashboardRedirection } from '@app/hooks';
import { Footer } from './Footer';
import { Header, HeaderLink, HeaderProps } from './Header';
import styles from './MarketingLayout.module.scss';

export interface MarketingLayoutProps {
  header?: React.ComponentType;
  children?: React.ReactNode;
  headerProps?: HeaderProps;
  hero?: boolean;
  dashboardRedirection?: boolean;
}

const links: HeaderLink[] = [
  { type: 'link', href: '/job-seekers/', label: 'Job Seekers' },
  { type: 'link', href: '/employers/', label: 'Employers' },
  { type: 'link', href: '/service-providers/', label: 'Service Providers' },
  { type: 'link', href: 'mailto:support@twout.org', label: 'Contact Us' },
];

export const MarketingLayout: React.FC<MarketingLayoutProps> = ({
  children,
  header: HeaderComponent = Header,
  headerProps,
  hero,
  dashboardRedirection,
}) => {
  const isLoading = useDashboardRedirection();

  return (
    <>
      <SkipNavLink>Skip to content</SkipNavLink>
      <HeaderComponent {...headerProps} links={links} isMarketingPage logoLink="/" />
      <SkipNavContent />
      <main
        className={classNames(
          hero && styles.hero,
          dashboardRedirection && isLoading && styles.isLoading,
        )}
      >
        {children}
      </main>
      <Footer />
    </>
  );
};
