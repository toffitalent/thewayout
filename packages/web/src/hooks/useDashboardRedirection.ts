import { useRouter } from 'next/router';
import { useEffect, useMemo, useState } from 'react';
import { useAppAuth, useAuthRedirect } from './useAppAuth';

export const useDashboardRedirection = () => {
  const router = useRouter();
  const auth = useAppAuth();
  const redirect = useAuthRedirect(false);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    if (auth.isLoading) {
      return;
    }
    if (auth.user && router.asPath === '/') {
      redirect();
    } else {
      setIsLoading(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [auth.user, auth.isLoading, router.asPath]);

  return useMemo(() => isLoading, [isLoading]);
};
