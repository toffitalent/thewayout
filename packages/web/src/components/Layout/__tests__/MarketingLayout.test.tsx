import { useRouter } from 'next/router';
import { useAppAuth, useDashboardRedirection } from '@app/hooks';
import { render, waitFor } from '@test';
import { MarketingLayout } from '../MarketingLayout';

jest.mock('@app/hooks', () => ({
  ...jest.requireActual('@app/hooks'),
  useAppAuth: jest.fn(),
  useDashboardRedirection: jest.fn(),
}));

describe('MarketingLayout', () => {
  beforeEach(() => {
    (useRouter as jest.Mock).mockImplementation(() => ({
      events: { on: jest.fn(), off: jest.fn() },
    }));
    (useAppAuth as any).mockImplementation(() => ({
      user: null,
      isLoading: false,
    }));
    (useDashboardRedirection as any).mockImplementation(() => false);
  });

  test('renders correctly', async () => {
    const { container, findByText } = render(
      <MarketingLayout>
        <div>Page</div>
      </MarketingLayout>,
    );
    await waitFor(async () => expect(await findByText('Job Seekers')).toBeInTheDocument());

    expect(container).toMatchSnapshot();
  });
});
