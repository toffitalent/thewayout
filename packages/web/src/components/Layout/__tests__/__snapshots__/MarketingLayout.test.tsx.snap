// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarketingLayout renders correctly 1`] = `
<div>
  <a
    class="skipNav"
    href="#skip-nav"
  >
    Skip to content
  </a>
  <header
    class="header"
  >
    <nav
      class="nav"
    >
      <a
        aria-label="Home"
        class="logo"
        href="/"
      />
      <button
        aria-controls="nav_links"
        aria-expanded="false"
        aria-label="Toggle navigation"
        class="toggle"
        type="button"
      >
        <div
          class="hamburger"
        >
          <div
            class="hamburgerInner"
          />
        </div>
      </button>
      <div
        class="wrapper"
      >
        <ul
          class="links"
          id="nav_links"
          role="menu"
        >
          <div
            class="renderedLinks"
          >
            <li
              role="menuitem"
            >
              <a
                class="link"
                href="/job-seekers"
              >
                Job Seekers
              </a>
            </li>
            <li
              role="menuitem"
            >
              <a
                class="link"
                href="/employers"
              >
                Employers
              </a>
            </li>
            <li
              role="menuitem"
            >
              <a
                class="link"
                href="/service-providers"
              >
                Service Providers
              </a>
            </li>
            <li
              role="menuitem"
            >
              <a
                class="link"
                href="mailto:support@twout.org"
              >
                Contact Us
              </a>
            </li>
          </div>
          <li
            role="menuitem"
          >
            <a
              class="link"
              href="/login"
            >
              Log In
            </a>
          </li>
          <li
            role="menuitem"
          >
            <a
              class="button md solid c-primary linkButton"
              href="/signup"
            >
              Sign Up
            </a>
          </li>
        </ul>
      </div>
    </nav>
  </header>
  <div
    id="skip-nav"
  />
  <main
    class=""
  >
    <div>
      Page
    </div>
  </main>
  <footer
    class="footer"
  >
    <div
      class="container"
    >
      <p
        class="text color-grey-500 fontWeight-700"
      >
        The Way Out, Inc. Â© 
        2024
        <a
          href="/terms"
        >
          Terms
        </a>
        <a
          href="/privacy"
        >
          Privacy
        </a>
      </p>
    </div>
  </footer>
</div>
`;
